Mrgiles1 timchang87/cdw 39 create onboarding for first time login

timchang87 opened the pull request:
Wasn't sure how bare the UI should be. So, this is just the skeleton with some conditional rendering at a field level. Some fields are shared between Mentor and Mentee (see ticket for the fields).  Upon clicking Submit, the page redirects to /home. Lastly, if reviewing this, to see the changes you will need to go to the /onboarding path after starting the frontend and pulling this branch. 

If this UI needs more meat, please let me know. This feature is great candidate for a multi-step or wizard form, and can be added, if needed.

timchang87 committed:
Pull prod

timchang87 committed:
Rebase prod

timchang87 committed:
Rebase prod

timchang87 committed:
Create Onboarding page and add a route (/onboarding)

timchang87 committed:
Add handleChange functionality and optimize conditional rendering to render at the field level not component level.

timchang87 committed:
Adjust formatting

jtabba commented on backend/build/api/v1/mentorShip/controllers/userController.d.ts:
```diff
@@ -0,0 +1,3 @@
+import { Request, Response } from 'express';
+export declare const getAll: (req: Request, res: Response) => Promise<Response<any, Record<string, any>>>;
+export declare const getSingle: (req: Request, res: Response) => Promise<Response<any, Record<string, any>>>;
```
Lets turn this into a custom interface without the use of any. The incoming type will be the data sent from the onboarding process

jtabba commented on frontend/src/App.tsx:
```diff
@@ -4,6 +4,7 @@ import { userData } from "./requests/types";
 import { UserInterface } from "./components/UserInterface";
 import { Routes, Route, BrowserRouter as Router } from "react-router-dom";
 import Home from "./pages/Home";
+import Onboarding from "./pages/Onboarding";
```
All pages are named using lowercase. An error may occur but will go away once you merge prod

timchang87 committed:
Add gitignore to backend.

jtabba committed:
.gitignore edit

jtabba committed:
.gitignore edit

jtabba commented on frontend/src/pages/Onboarding.tsx:
```diff
@@ -0,0 +1,172 @@
+        jobTitle: '',
+        interests: '',
+        hourlyRate: ''
+    });
```
Create an interface for this data

timchang87 replied:
Ended up create additional states for Mentor and Mentee and changing the existing state to track which role was chosen in the initial question. Created an interface for each.

jtabba commented on frontend/src/pages/Onboarding.tsx:
```diff
@@ -0,0 +1,172 @@
+
+    const handleChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement >) => {
+        setUserInfo({...userInfo, [e.target.name]: e.target.value})
+    };
```
Excellent

put a space between the `{ ... }`

timchang87 replied:
Formatting has been changed.

jtabba commented on frontend/src/pages/Onboarding.tsx:
```diff
@@ -0,0 +1,172 @@
+                    p={5}
+                >
+      
+                        <Heading>{(role === 'mentee' && 'Mentee') || (role === 'mentor' && 'Mentor')}</Heading>
```
Lets turn this into one statement using a ternary

timchang87 replied:
This has been removed as separate UIs were created for Mentor and Mentee.

jtabba reviewed (changes requested):
Great job overall. A few suggestions to polish this to what would be considered a high standard:
- Theres a lot of reused `Flex` components. See if you can group the common properties to create a custom component using `forwardProp` from chakra. Will knock down the line count/clutter by a lot (bonus points if you do it for other components)
- Conditional rendering: I can see that you have conditionally rendered the different inputs based on whether the user is a mentor or mentee. Not 'bad' but it can be improved for better legibility if we create an if statement that renders either a UI for the mentor questions or mentee questions. If you really want to get fancy you can create a component that takes in the data state and maps over it to render the input based on what is available in the data set.
- Positioning: When loaded the question box is pretty high up the page. Lets center it a bit more

mrgiles1 committed:
tried to edit theme, ultimated added Logo.css to implement font

mrgiles1 committed:
got custom font working via theme.ts

mrgiles1 committed:
added logo to Hero.tsx, edited headline for responsiveness testing

timchang87 committed:
Pull prod

timchang87 committed:
Make changes from code review feedback
Create FormContainer and FormRow components
Move Mentee/Mentor question down
Create separate UIs for mentor and mentee questions that conditionally render using if statement
Make formatting changes
Make react page filenames lowercase

timchang87 committed:
Fix merge conflicts.

timchang87 commented:
- Ended up creating new components to group the `Flex` components with similar props. However, I think I may have done a little more than necessary. See FormRow and FormContainer. It feels like overkill but would like to get your thoughts.
- Conditional rendering has been changed. I used a function and an `if` statement to render the appropriate components.
- Fixed positioning
- Lastly, I tried to make the page more responsive and implemented responsive styling. Needs some tweaking but for the most part seems to be working.

timchang87 committed:
Fix formating

timchang87 committed:
Remove build folder.

mrgiles1 committed:
update profile page to get user picture, name, email, and username from AUTH0

timchang87 committed:
Create separate components for Mentee and Mentor onboarding.

timchang87 commented:
- Created separate components for Mentor and Mentee onboarding forms (MentorForm and MenteeForm). I do think there is an opportunity here to create a generic OnboardingForm component that we could reuse for both Mentor and Mentee forms. However, I'm lacking the brain power to think through this today. And I think I will have to have a better understanding of generics. 

- Part of the process of creating separate components, I've moved state for menteeInfo and mentorInfo into their respective components. The state does not need to be shared by both components at the Onboarding level. So I thought moving them down to MenteeForm and MentorForm was better practice.

- Instead of using window.location.href, I've imported react router and am using useNavigate.

timchang87 committed:
Fix issue with lowercase changes not pushing to GitHub.

jtabba committed:
Merge branch 'prod' into mrgiles1-timchang87/cdw-39-create-onboarding-for-first-time-login

jtabba commented on frontend/src/components/MentorForm.tsx:
```diff
@@ -0,0 +1,79 @@
+
+interface MentorFields {
+    mentorFields: Field[] 
+}
```
We've got some duplication happening here. Export fields into a shared file and then make MentoreFields/MenteeFields a type which = Field (rather than an interface as we don't intend on extending them)

Aside from that Mentor/MenteeInfo can be created into a base interface which then extends the required differing fields. That way if we need to add more to the form we only need to do it in one place

jtabba commented on frontend/src/pages/onboarding.tsx:
```diff
@@ -0,0 +1,98 @@
+      name: 'hourlyRate',
+    },
+  ];
+
```
Export all of the above into a `constants.ts` file (put in src root) and change casing to UPPER_CASE. Can also create one variable for shared attributes and then extend the required ones for mentee/mentor

timchang87 replied:
The following has been completed:

- Created constants.tsx
- Created one variable called QUESTIONNAIRE_FIELDS for the shared attributes and extended this variable for each mentee and mentor variable, spreading the QUESTIONNAIRE_FIELDS into each.
- Exported the mentor and mentee variables and imported those into onbaording.tsx and referenced those variables when passing props to the respective child components.
- Deleted the mentee and mentor variables on onboarding.tsx

timchang87 committed:
Optimize constants used in onboarding to be in separate global file.

timchang87 committed:
Optimize types for onboarding form.

timchang87 committed:
Merge branch 'prod' into mrgiles1-timchang87/cdw-39-create-onboarding-for-first-time-login
